File: concepts/10_time_locale_and_runtime/01_durations_clocks_and_monotonic_time/cpp/test_02_units_and_clock_boundaries.cpp
// 时长单位和时钟边界。
// 共同问题：时长是否携带单位；墙上时间回拨时经过时间如何计算；
// 数值精度与单位转换由调用方还是类型系统约束。
//
// polyglot-family: time_locale_and_runtime
// polyglot-concept: durations_clocks_and_monotonic_time
// polyglot-related: languages/cpp/standard_library/12_time/
// polyglot-related+: test_110_duration_period_conversions_arithmetic_rounding_and_literals.cpp

#include <gtest/gtest.h>

#include <chrono>
#include <type_traits>

namespace {

using namespace std::chrono_literals;

TEST(DurationBoundaryConcept, CommonTypePreservesTheFinerUnit) {
  const auto duration = 1s + 250ms;

  static_assert(std::is_same_v<decltype(duration), const std::chrono::milliseconds>);
  EXPECT_EQ(duration.count(), 1'250);
  EXPECT_EQ(std::chrono::duration_cast<std::chrono::seconds>(duration), 1s);
}

TEST(DurationBoundaryConcept, IntegralCastTruncatesTowardZero) {
  const auto positive = std::chrono::duration_cast<std::chrono::seconds>(1'900ms);
  const auto negative = std::chrono::duration_cast<std::chrono::seconds>(-1'900ms);

  EXPECT_EQ(positive, 1s);
  EXPECT_EQ(negative, -1s);
}

TEST(DurationBoundaryConcept, ClockChoiceDeterminesWhetherRollbackLooksLikeElapsedTime) {
  const std::chrono::system_clock::time_point wall_start{1'000s};
  const std::chrono::system_clock::time_point wall_finish{995s};
  const std::chrono::steady_clock::time_point steady_start{40s};
  const std::chrono::steady_clock::time_point steady_finish{42s + 500ms};

  EXPECT_EQ(wall_finish - wall_start, -5s);
  EXPECT_EQ(steady_finish - steady_start, 2s + 500ms);

  // 注入时间点避免依赖真实校时；不同 clock 的 time_point 不能直接混算。
}

}  // namespace

File: concepts/10_time_locale_and_runtime/01_durations_clocks_and_monotonic_time/julia/test_02_units_and_clock_boundaries.jl
# polyglot-family: time_locale_and_runtime
# polyglot-concept: durations_clocks_and_monotonic_time
# polyglot-related: languages/julia/standard_library/09_text_data_dates_and_math/
# polyglot-related+: test_069_dates_periods_and_calendar_arithmetic.jl
#
# 共同问题：不同 duration 单位是否隐式混为裸数字；calendar period 能否当固定秒数。
# 对照观察：Dates period 在类型中保留单位；Month 是日历步进，不等于固定数量的 Millisecond。

using Test
using Dates

@testset "固定单位与 calendar unit 不伪装等价" begin
    @test canonicalize(Millisecond(120_000)) == Minute(2)
    @test Dates.value(Hour(3)) == 3
    @test Hour(1) != Millisecond(1)
    @test typeof(Month(1)) === Month
    @test_throws MethodError convert(Millisecond, Month(1))
end
